### PR TITLE
Update Imaginary to optionally process files locally

### LINKED ISSUE #44708: Add support for Imaginary to generate previews via "file" parameter through locally mounted folder with "-mount" command line option
The current implementation of Imaginary preview generation in Nextcloud works by submitting the image via a POST to the configured 'preview_imaginary_url' in config.php.

While this works and is scalable, I believe on certain (and possibly many) configurations, this is the slower route wherein the Imaginary server (whether running on baremetal, a jail, or Docker container) may have local file access to the image. Thus, having NC server send the entire image via a POST to Imaginary server is an unnecessary use of bandwidth and has network overhead, and is slower than local file access.

Imaginary currently has the ability to instead use a "file" parameter via the POST request to direct Imaginary to generate a preview for a file that it has local access to via the "-mount" parameter provided on the command line of Imaginary:
https://github.com/h2non/imaginary

> - file string - Use image from server local file path. In order to use this you must pass the -mount=<dir> flag.

**Describe the solution you'd like**
A boolean option in config.php such as 'preview_imaginary_local_mount' that would direct NC to instead send a proper file path via "file" and not upload the entire file to Imaginary.  This would, of course, necessitate that the instance of Imaginary being used is configured properly with the NC Data folder mounted.

**Describe alternatives you've considered**
ImageMagick previews are currently generated with direct access to the files due to the fact that ImageMagick runs on the same server via a PHP Module.  I recently switched over to Imaginary because of 1) webp previews being possible and 2) slight security issues in ImageMagick (none that I've encountered, but I've been urged to jump to Imaginary.) I would sleep better at night (lol) knowing that Imaginary is running as fast as or faster than ImageMagick for preview generation but as long as NC has to send an entire file via POST to Imaginary, I think that would cause a large performance hit via a locally access ImageMagick generated preview.

If I have some time, I may run some tests with say 100 random images in my NC instance and use PreviewGenerator to preview:generate-all those 100 images with Imaginary and ImageMagick for comparison.

Thank you for your time if you read this.

### PR DESCRIPTION
**Pull request by invario**
Enhancement proposal from my request:
https://github.com/nextcloud/server/issues/44708

Updates Imaginary to look for a new configuration parameter 'preview_imaginary_locally_mounted" in config.php.  If set to 'true', it will pass the file path to Imaginary to process locally.  If set to 'false', it will process it the existing way by uploading the file via POST.

The Imaginary server must have been initiated with "-mount /foo/bar/NEXTCLOUD-DATA-DIRECTORY" in order to function properly.

* Resolves: #44708  

## TODO

- [ ] ...

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [ ] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [ ] Screenshots before/after for front-end changes
- [ ] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [ ] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

### CODE REVIEW

**Review by invario — COMMENTED** [review]
Apologies, I'm new at this. FIxed getSystemValueString to use Bool instead

File: lib/private/Preview/Imaginary.php [hunk 1]
```diff
@@ -76,12 +76,15 @@
 		}
 		$imaginaryUrl = rtrim($imaginaryUrl, '/');
 
-		// Object store
-		$stream = $file->fopen('r');
-		if (!$stream || !is_resource($stream) || feof($stream)) {
-			return null;
+		// Set to true if Imaginary was configured with "-mount" with the NextCloud data directory, else set to false
+		$locallymounted = $this->config->getSystemValueString('preview_imaginary_locally_mounted',false);
```
[the remark this answers is not in the document]
Comment (invario) [reply lib/private/Preview/Imaginary.php]: Fixed in latest

**Comment by solracsf** [comment]
Thanks for you work, but regarding https://github.com/nextcloud/server/issues/44708#issuecomment-2043853963 I wonder if this is "that" valuable to be implemented 🤔

**Comment by invario** [comment]
> Thanks for you work, but regarding [#44708 (comment)](https://github.com/nextcloud/server/issues/44708#issuecomment-2043853963) I wonder if this is "that" valuable to be implemented 🤔

I ran some tests and posted them here:
https://github.com/nextcloud/server/issues/44708#issuecomment-2043853963
Which show on my own specific setup, it does **_not_** improve performance when it comes to preview generation time. However, I would caution everyone not to conclude that for all use cases there wouldn't be an impact on performance. I also did not measure other factors like average LOAD as a result of processing a file upload via POST versus locally accessing a file.

I can see there being a difference if things were configured differently with an Imaginary server separate from the NC server with both utilizing the same NC Data directory via network.

In the existing case mode, the following must be done to generate a preview:

1. NC opens file and POSTs to Imaginary with entire file and requested image operations
2. Imaginary waits for the entire file to be transferred
3. Imaginary opens the file
4. Imaginary generates the preview
5. Imaginary returns the image to NC

With my proposed local access option, it would be reduced to:

1. NC submits GET to Imaginary with requested image operations
2. Imaginary opens the file direct from storage
4. Imaginary generates the preview
5. Imaginary returns the image to NC

At the very least, I think we would save a fopen on the NC server.

IMHO, I don't think it _**hurts**_ to implement this as a user configurable option. It only requires minimal changes and would simply default to how it's always run.

Thank you for your time! :)

**Comment by kesselb** [comment]
Does it work with an external storage?

The datadirectory is still local and there is also an external storage configured. 

I would expect a couple of more switches to distinguish if the file is expected to be in the data directory or remote and therefore needs the other way then.

**Comment by invario** [comment]
> Does it work with an external storage?
> 
> The datadirectory is still local and there is also an external storage configured.
> 
> I would expect a couple of more switches to distinguish if the file is expected to be in the data directory or remote and therefore needs the other way then.

That is a good point and good catch. I did not test for that at all.  Back to the testing board for me.

**Comment by invario** [comment]
> > At the very least, I think we would save a fopen on the NC server.
> 
> https://stackify.com/premature-optimization-evil/ ;)

Thank you. As I said, I'm very new to this and trying to learn. :)

**Comment by invario** [comment]
Closed, my proposed implementation doesn't work for external storage.

**Pull request closed without merging**